### ISSUE TITLE
Interfaces with fields that are interfaces

### ISSUE THREAD

**Issue by sdcooke** [graphql-python/graphql-core-legacy#48]
Apologies if this is a little convoluted - it's hard to explain!

https://github.com/facebook/graphql/pull/123 allows implementations of interfaces to have fields that are implementations of interfaces. Currently `assert_object_implements_interface` (`graphql/core/type/schema.py`) only allows the fields to match types exactly. An example:

```
class InterfaceA(graphene.Interface):
    ...

class InterfaceB(graphene.Interface):
    my_field = graphene.List(InterfaceA)
    ...

class ImplementationA(InterfaceA):
    ...

class ImplementationB(InterfaceB):
    my_field = graphene.List(ImplementationA)
    ...
```

To test something out I made `assert_object_implements_interface` a no-op then hit further issues with `graphql/core/validation/rules/overlapping_fields_can_be_merged.py`. A simplified part of a query (in practise the query was built up from fragments coming from different places) returning an object of type ImplementationB:

```
my_field
... on InterfaceB {
    my_field
}
```

In theory it should be able to merge these, but in practice the first `my_field` is known to be `ImplementationA` and the second only `InterfaceA`.

I think the first part of this issue is worth fixing and I think I could probably manage that and open a PR if you'd like.

The second part I'm less sure about and I would have a harder time fixing - let me know what you think though. It's possible to work around the second problem by making sure queries are wrapped with `... on InterfaceB` when merging might happen.

**Comment by syrusakbary** [comment]
@sdcooke this will be implemented quite soon. Thanks for the detailed issue!

**Comment by syrusakbary** [comment]
This is implemented in the current version `0.5.0`, closing the issue.
